src/test/java/unit/DataSourceReferenced.java

```diff
@@ -0,0 +1,111 @@
+package unit;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import bio.terra.cli.businessobject.resource.DataSource;
+import bio.terra.cli.serialization.userfacing.UFWorkspace;
+import bio.terra.cli.serialization.userfacing.resource.UFDataSource;
+import bio.terra.cli.service.WorkspaceManagerService;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.common.collect.MoreCollectors;
+import harness.TestCommand;
+import harness.baseclasses.SingleWorkspaceUnit;
+import harness.utils.TestUtils;
+import harness.utils.WorkspaceUtils;
+import java.io.IOException;
+import java.util.List;
+import java.util.UUID;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+/** Tests for the `terra resource` commands that handle data sources. */
+@Tag("unit")
+public class DataSourceReferenced extends SingleWorkspaceUnit {
+  private static final String thousandGenomesWorkspaceName = "1000 Genomes";
+  private static final String thousandGenomesShortDescription = "short description";
+  private static final String thousandGenomesVersion = "version";
+  private static final String thousandGenomesDescription = "description";
+  private static final String thousandGenomesResourceName =
+      TestUtils.appendRandomNumber("1000-genomes-ref");
+  private UUID thousandGenomesUuid;
+
+  @BeforeAll
+  @Override
+  protected void setupOnce() throws Exception {
+    super.setupOnce();
+
+    // Set up 1000 Genomes data source workspace
+    String properties =
+        String.format(
+            "%s=%s,%s=%s",
+            DataSource.SHORT_DESCRIPTION_KEY,
+            thousandGenomesShortDescription,
+            DataSource.VERSION_KEY,
+            thousandGenomesVersion);
+    UFWorkspace thousandGenomesWorkspace =
+        WorkspaceUtils.createWorkspace(
+            workspaceCreator, thousandGenomesWorkspaceName, thousandGenomesDescription, properties);
+
+    // Add 1000 Genomes data source to researcher workspace. This isn't supported by CLI, so call
+    // WSM directly.
+    UUID researcherWorkspaceUuid =
+        WorkspaceManagerService.fromContext().getWorkspaceByUserFacingId(getUserFacingId()).getId();
+    thousandGenomesUuid =
+        WorkspaceManagerService.fromContext()
+            .getWorkspaceByUserFacingId(thousandGenomesWorkspace.id)
+            .getId();
+    WorkspaceManagerService.fromContext()
+        .createReferencedTerraWorkspace(
```

Comment:
just do terra resource add-ref datasource instead and remove the test-only method.

Reply:
i see your point. Is the plan to not support adding a datasource through cli? it doesn't feel symmetrical that you can update delete datasource, but not create through CLI.

Reply:
Yeah, delete didn't have the problems that add-ref had. And it was easy to implement. It is a strange that it's not symmetrical.